packages/flutter/lib/src/services/spell_check.dart

```diff
@@ -0,0 +1,553 @@
+       String oldSpanText;
+       String newSpanText;
+       int spanLength = 0;
+       int newStart = 0;
+       int searchStart = 0;
+
+       int? start_index;
+       int? end_index;
+       bool spanWithinTextRange = true;
+
+       int currentSpanStart = 0;
+       int currentSpanEnd = 0;
+
+       while(span_pointer < results.length) {
+           currentSpan = results[span_pointer];
+           currentSpanStart = currentSpan.start;
+           currentSpanEnd = currentSpan.end;
+
+           start_index = currentSpanStart < newText.length ? currentSpanStart : null;
+           end_index = currentSpanEnd < newText.length ? currentSpanEnd : null;
+
+           spanWithinTextRange = start_index != null && end_index != null;
+
+           if (!spanWithinTextRange) {
+               // No more of the spell check results will be within the range of the text
+               break;
+           } 
+           else {
+              oldSpanText = resultsText.substring(currentSpanStart, currentSpanEnd + 1);
+              newSpanText = newText.substring(currentSpanStart, currentSpanEnd + 1);
+
+                if (oldSpanText == newSpanText) {
+                    searchStart = currentSpanEnd + 1;
+                    correctedSpellCheckResults.add(currentSpan);
+                }
+                else {
+                    spanLength = currentSpanEnd - currentSpanStart; 
+                    RegExp regex = RegExp('\\b$oldSpanText\\b');
+                    int substring = newText.substring(searchStart).indexOf(regex);
+                    newStart = substring + searchStart;
+
+                    if (substring >= 0) {
+                        correctedSpellCheckResults.add(SpellCheckerSuggestionSpan(newStart, newStart + spanLength, currentSpan.replacementSuggestions));
+                        searchStart = newStart + spanLength;
+                    }
+                }
+           }
+
+           span_pointer += 1;
+       }
+
+       return correctedSpellCheckResults;
+   }
+
+   // pretty sure this can be replaced with a String method
+   int? findBadSpan(String text, String spanText, int spanLength) {
+       bool foundSpan = false;
+       int text_pointer = 0;
+
+       while (!foundSpan && text_pointer + spanLength < text.length) {
+           int end = text_pointer + spanText.length;
+
+           if (text.substring(text_pointer, end) == spanText) {
+               return text_pointer;
+           }
+
+           text_pointer += 1;
+       }
+     return null;
+   }     
+
+  @override
+  TextSpan buildTextSpanWithSpellCheckSuggestions(
+      List<SpellCheckerSuggestionSpan>? rawSpellCheckResults, String? spellCheckResultsText,
+      TextEditingValue value, TextStyle? style, bool composingWithinCurrentTextRange) {
+      scssSpans_consumed_index = 0;
+      text_consumed_index = 0;
+
+      List<SpellCheckerSuggestionSpan>? spellCheckResults;
+
+      if (spellCheckResultsText != value.text) {
```

Comment:
What if they're not equal, but spellCheckResultsText is a substring of value.text?  That is probably a very common scenario, and you could save some time by not doing the full correctSpellCheckResults in that case.